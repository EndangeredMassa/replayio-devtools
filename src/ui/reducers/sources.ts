import {
  createEntityAdapter,
  createSelector,
  createSlice,
  EntityState,
  PayloadAction,
} from "@reduxjs/toolkit";
import { newSource, SourceKind } from "@replayio/protocol";
import uniq from "lodash/uniq";
import omit from "lodash/omit";
import { getSelectedSourceId } from "devtools/client/debugger/src/selectors";
import { UIState } from "ui/state";
// sourceId: SourceId;
// kind: SourceKind;
// url?: string;
// generatedSourceIds?: SourceId[];
// contentHash?: string;

interface SourceDetails {
  contentHash?: string;
  id: string;
  kind: SourceKind;
  url: string;
}
const sourceDetailsAdapter = createEntityAdapter<SourceDetails>();
const sourcesAdapter = createEntityAdapter<newSource>({ selectId: source => source.sourceId });

export interface SourcesState {
  generated: { [key: string]: string[] | undefined };
  generatedFrom: { [key: string]: string[] | undefined };
  prettyPrintedFrom: { [key: string]: string | undefined };
  prettyPrintedTo: { [key: string]: string | undefined };
  sourceDetails: EntityState<SourceDetails>;
  sources: EntityState<newSource>;
}

const initialState: SourcesState = {
  generated: {},
  generatedFrom: {},
  prettyPrintedFrom: {},
  prettyPrintedTo: {},
  sourceDetails: sourceDetailsAdapter.getInitialState(),
  sources: sourcesAdapter.getInitialState(),
};

const sourcesSlice = createSlice({
  name: "sources",
  initialState,
  reducers: {
    addSource: (state, action: PayloadAction<newSource>) => {
      // Always keep the raw protocol information
      sourcesAdapter.addOne(state.sources, action.payload);

      if (action.payload.kind === "prettyPrinted") {
        const id = action.payload.sourceId;
        // This is confusing because we turn things around when pretty printing
        const prettyPrintedFrom = action.payload.generatedSourceIds![0]!;
        // Unclear to me if we should consider pretty-printing to be a source transformation?
        // state.generated[id] = uniq([...(state.generated[id] || []), prettyPrintedFrom]);
        state.prettyPrintedFrom[id] = prettyPrintedFrom;
        state.prettyPrintedTo[prettyPrintedFrom] = id;

        sourceDetailsAdapter.upsertOne(state.sourceDetails, {
          ...omit(action.payload, "sourceId", "generatedSourceIds"),
          id: prettyPrintedFrom,
          // This source was pretty printed, and (I think) therefore not
          // generated from eval, so we can safely assume that we have a url...
          // for now at least.
          url: action.payload.url!,
        });
      } else {
        const id = action.payload.sourceId;
        if (action.payload.generatedSourceIds) {
          // Link this to the sources it generated
          state.generated[id] = uniq([
            ...(state.generated[id] || []),
            ...action.payload.generatedSourceIds,
          ]);
          // Link each source it generated back to this
          action.payload.generatedSourceIds.map(generatedSourceId => {
            state.generatedFrom[generatedSourceId] = uniq([
              ...(state.generatedFrom[generatedSourceId] || []),
              id,
            ]);
          });
        }
        sourceDetailsAdapter.upsertOne(state.sourceDetails, {
          ...omit(action.payload, "sourceId", "generatedSourceIds"),
          id: action.payload.sourceId,
          // I don't have a good excuse as to why I am saying this exists even
          // though it is optional. Other than - we'll get to that later.
          url: action.payload.url!,
        });
      }
    },
  },
});

const sourceDetailsSelectors = sourceDetailsAdapter.getSelectors();

export const getSelectedSourceDetails = createSelector(
  getSelectedSourceId,
  (state: UIState) => state.newSources.sourceDetails,
  (state: UIState) => state.newSources.generated,
  (state: UIState) => state.newSources.generatedFrom,
  (state: UIState) => state.newSources.prettyPrintedFrom,
  (state: UIState) => state.newSources.prettyPrintedTo,
  (
    selectedSourceId,
    sourceDetails,
    generated,
    generatedFrom,
    prettyPrintedFrom,
    prettyPrintedTo
  ) => {
    if (!selectedSourceId) {
      return null;
    }
    const beforePrettyPrinting = prettyPrintedFrom[selectedSourceId];
    const details = sourceDetailsSelectors.selectById(
      sourceDetails,
      beforePrettyPrinting || selectedSourceId
    );

    if (!details) {
      return null;
    }

    let canonicalSource;
    if (beforePrettyPrinting !== undefined) {
      // For the purposes of demonstration, let's say that prettyPrinted files
      // are the canonical ones already. I have not yet confirmed if this is
      // reliably the case, and I suspect that it is not. In fact, I suspect the
      // logic is:
      // - Go find the canonical source for the non-pretty-printed version, and
      //   then see if that canonical source also has a pretty-printed version.
      canonicalSource = null;
    } else if (details.kind === "html") {
      // HTML files are always the original, canonical source AFAIK
      canonicalSource = null;
    } else if (details.kind === "inlineScript") {
      // HTML files are the canonical source for inline scripts
      canonicalSource = generatedFrom[selectedSourceId]![0] || null;
    } else if (details.kind === "scriptSource") {
      // This might be a file referenced directly, but if there's a pretty-printed version, we would prefer that.
      canonicalSource = prettyPrintedTo[selectedSourceId] || null;
    } else if (details.kind === "sourceMapped") {
      // This might be a sourceMapped file, but we still want the PP version if possible
      canonicalSource = prettyPrintedTo[selectedSourceId] || null;
    } else if (details.kind === "other") {
      // I don't know what to do here yet, probably generatedFrom?
      canonicalSource = generatedFrom[selectedSourceId]![0] || null;
    }

    return {
      ...details,
      canonicalSource,
      generated: generated[selectedSourceId] || [],
      generatedFrom: generatedFrom[selectedSourceId] || [],
      prettyPrintedFrom: prettyPrintedFrom[selectedSourceId],
      prettyPrintedTo: prettyPrintedTo[selectedSourceId],
    };
  }
);

export const { addSource } = sourcesSlice.actions;
export default sourcesSlice.reducer;
