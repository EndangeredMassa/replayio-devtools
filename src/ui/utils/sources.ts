import { EntityId } from "@reduxjs/toolkit";
import { newSource, SourceKind } from "@replayio/protocol";
import groupBy from "lodash/groupBy";
import { SourceDetails } from "ui/reducers/sources";

const fullSourceDetails = (
  attributes: Partial<SourceDetails> & {
    id: string;
    kind: SourceKind;
    url: string;
  }
): SourceDetails => {
  return {
    canonicalId: attributes.id,
    contentHash: undefined,
    correspondingSourceIds: [],
    generated: [],
    generatedFrom: [],
    prettyPrinted: undefined,
    prettyPrintedFrom: undefined,
    ...attributes,
  };
};

export const keyForSource = (source: newSource): string => {
  return `${source.url!}:${source.contentHash}`;
};

export const newSourcesToCompleteSourceDetails = (
  newSources: newSource[]
): Record<EntityId, SourceDetails> => {
  // Sources are processed by kind. So first we go through the whole list once
  // just to group things properly.
  const byKind = groupBy(newSources, source => source.kind);

  // Process scriptSources first. Generally speaking, all other resources will
  // be related to a scriptSource *somehow*. There are exceptions. For instance,
  // the pretty-printed version of a sourceMapped source will not have a direct
  // link to a scriptSource (though it must have a transitive link through the
  // sourceMapped source)
  const scriptSources = byKind["scriptSource"] || [];

  const returnValue: Record<EntityId, SourceDetails> = {};

  const generatedFromMap: Record<string, string[] | undefined> = {};
  const backLinkGeneratedSource = (source: newSource) => {
    source.generatedSourceIds?.map(generatedId => {
      if (!generatedFromMap[generatedId]) {
        generatedFromMap[generatedId] = [];
      }
      generatedFromMap[generatedId]!.push(source.sourceId);
    });
  };

  // Rather than searching by the canonicalID, we can use this map to do a
  // reverse lookup.
  const canonicalSourcesMap: Record<string, string[] | undefined> = {};
  const addCanonicalLink = (replace: string, replaceWith: string) => {
    const existingCanonical = canonicalSourcesMap[replace] || [];
    existingCanonical.forEach(sourceId => {
      returnValue[sourceId]!.canonicalId = replaceWith;
    });
    console.log(`Replace ${replace}(${existingCanonical}) -> ${replaceWith}`);
    canonicalSourcesMap[replaceWith] = [...existingCanonical, replaceWith];
    if (replace !== replaceWith) {
      delete canonicalSourcesMap[replace];
    }
  };

  const lookupCanonicalSourceId = (sourceId: string) => {
    let candidate = sourceId;

    while (returnValue[candidate] && returnValue[candidate].canonicalId !== candidate) {
      // Keep following the links
      candidate = returnValue[candidate].canonicalId;
    }

    console.log(`Lookup ${sourceId} => ${candidate}`);
    return candidate;
  };

  const correspondingSourcesMap: Record<string, string[] | undefined> = {};
  const addToCorrespondingSources = (source: newSource) => {
    const key = keyForSource(source);
    if (!correspondingSourcesMap[key]) {
      correspondingSourcesMap[key] = [];
    }
    correspondingSourcesMap[key]!.push(source.sourceId);
  };

  scriptSources.map(source => {
    returnValue[source.sourceId] = fullSourceDetails({
      contentHash: source.contentHash!,
      generated: source.generatedSourceIds || [],
      id: source.sourceId,
      kind: source.kind,
      url: source.url!,
    });

    backLinkGeneratedSource(source);
    addToCorrespondingSources(source);
    canonicalSourcesMap[source.sourceId] = [source.sourceId];
  });

  const htmlSources = byKind["html"] || [];
  htmlSources.map(source => {
    returnValue[source.sourceId] = fullSourceDetails({
      contentHash: source.contentHash!,
      generated: source.generatedSourceIds || [],
      generatedFrom: generatedFromMap[source.sourceId] || [],
      id: source.sourceId,
      kind: source.kind,
      url: source.url!,
    });

    backLinkGeneratedSource(source);
    addToCorrespondingSources(source);
    addCanonicalLink(source.sourceId, source.sourceId);
  });

  const inlineScripts = byKind["inlineScript"] || [];
  inlineScripts.map(source => {
    const canonicalId = lookupCanonicalSourceId(generatedFromMap[source.sourceId]![0]);
    returnValue[source.sourceId] = fullSourceDetails({
      canonicalId,
      contentHash: source.contentHash,
      generated: source.generatedSourceIds || [],
      generatedFrom: generatedFromMap[source.sourceId] || [],
      id: source.sourceId,
      kind: source.kind,
      url: source.url!,
    });

    source.generatedSourceIds?.map(generatedId => {
      returnValue[generatedId]!.generatedFrom.push(source.sourceId);
    });

    backLinkGeneratedSource(source);
    addToCorrespondingSources(source);
    addCanonicalLink(source.sourceId, canonicalId);
  });

  const sourceMapped = byKind["sourceMapped"] || [];
  sourceMapped.map(source => {
    // If this source was source-mapped, then it must be the canonical source
    // for the things it generated?

    const canonicalId = lookupCanonicalSourceId(source.sourceId);
    returnValue[source.sourceId] = fullSourceDetails({
      canonicalId,
      contentHash: source.contentHash,
      generated: source.generatedSourceIds || [],
      generatedFrom: generatedFromMap[source.sourceId] || [],
      id: source.sourceId,
      kind: source.kind,
      url: source.url!,
    });

    // Link generated sources
    source.generatedSourceIds?.map(generatedId => {
      returnValue[generatedId]!.generatedFrom.push(source.sourceId);
      addCanonicalLink(generatedId, canonicalId);
    });

    backLinkGeneratedSource(source);
    addToCorrespondingSources(source);
  });

  const otherSources = byKind["other"] || [];
  otherSources.map(source => {
    returnValue[source.sourceId] = fullSourceDetails({
      generated: source.generatedSourceIds || [],
      generatedFrom: generatedFromMap[source.sourceId] || [],
      id: source.sourceId,
      kind: source.kind,
      url: source.url!,
    });

    // Link generated sources
    source.generatedSourceIds?.map(generatedId => {
      returnValue[generatedId]!.generatedFrom.push(source.sourceId);
    });

    backLinkGeneratedSource(source);
    addToCorrespondingSources(source);
  });

  const prettyPrinted = byKind["prettyPrinted"] || [];
  prettyPrinted.map(source => {
    // We handle pretty-printed (pp) files and their generated links a little
    // differently. Because Replay makes the pp sources, their structure is
    // predictable. All pp sources will have one generatedSourceId, and it will
    // be the minified source.
    const nonPrettyPrintedVersion = returnValue[source.generatedSourceIds![0]];
    const canonicalId = lookupCanonicalSourceId(nonPrettyPrintedVersion.canonicalId);
    returnValue[source.sourceId] = fullSourceDetails({
      canonicalId,
      id: source.sourceId,
      kind: source.kind,
      prettyPrintedFrom: source.generatedSourceIds![0]!,
      generatedFrom: generatedFromMap[source.generatedSourceIds![0]] || [],
      url: source.url!,
    });

    console.log(`Handling backlinks for pretty-printed ${source.sourceId}`);
    nonPrettyPrintedVersion.prettyPrinted = source.sourceId;
    addToCorrespondingSources(source);
    addCanonicalLink(nonPrettyPrintedVersion.id, source.sourceId);
  });

  // if we see an evaled script or an excerpted script we can mark it as
  // transitive and call the protocol to get its mapped location for first and
  // last location and BOOM!
  // Oh, ya know what, you only need *one* location because the source will
  // always be the same!
  // Ah, drat, for some reason eval'ed sources in particular are not linked!
  // That is not a huge problem right now, but eventually it *could* be.

  return returnValue;
};
