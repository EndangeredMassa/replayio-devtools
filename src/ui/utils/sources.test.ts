import { newSource } from "@replayio/protocol";
import { ThreadFront } from "protocol/thread";
import { newSourcesToCompleteSourceDetails } from "./sources";

describe("newSourcesToCompleteSourceDetails", () => {
  it("should return empty array when no sources", () => {
    expect(newSourcesToCompleteSourceDetails([])).toEqual({});
  });

  it("should be able to return a complete source from a single newSource event", () => {
    expect(
      newSourcesToCompleteSourceDetails([
        {
          contentHash: "contentHash#1",
          kind: "scriptSource",
          sourceId: "1",
          url: "/index.js",
        },
      ])
    ).toEqual({
      "1": {
        canonicalId: "1",
        contentHash: "contentHash#1",
        correspondingSourceIds: ["1"],
        generated: [],
        generatedFrom: [],
        id: "1",
        kind: "scriptSource",
        prettyPrinted: undefined,
        prettyPrintedFrom: undefined,
        url: "/index.js",
      },
    });
  });

  it("can backlink generated sources properly", () => {
    expect(
      newSourcesToCompleteSourceDetails([
        {
          contentHash: "contentHash#1",
          kind: "scriptSource",
          sourceId: "1",
          url: "/index.js",
        },
        {
          contentHash: "contentHash#o1",
          kind: "sourceMapped",
          sourceId: "o1",
          generatedSourceIds: ["1"],
          url: "/index.ts",
        },
      ])
    ).toEqual({
      "1": {
        canonicalId: "o1",
        contentHash: "contentHash#1",
        correspondingSourceIds: ["1"],
        generated: [],
        generatedFrom: ["o1"],
        id: "1",
        kind: "scriptSource",
        prettyPrinted: undefined,
        prettyPrintedFrom: undefined,
        url: "/index.js",
      },
      o1: {
        canonicalId: "o1",
        contentHash: "contentHash#o1",
        correspondingSourceIds: ["o1"],
        generated: ["1"],
        generatedFrom: [],
        id: "o1",
        kind: "sourceMapped",
        prettyPrinted: undefined,
        prettyPrintedFrom: undefined,
        url: "/index.ts",
      },
    });
  });

  it("can link pretty-printed and minified sources", () => {
    expect(
      newSourcesToCompleteSourceDetails([
        {
          contentHash: "contentHash#1",
          kind: "scriptSource",
          sourceId: "1",
          url: "/index.js",
        },
        {
          generatedSourceIds: ["1"],
          kind: "prettyPrinted",
          sourceId: "pp1",
          url: "/src/index.js",
        },
      ])
    ).toEqual({
      "1": {
        canonicalId: "1",
        contentHash: "contentHash#1",
        correspondingSourceIds: ["1"],
        generated: [],
        generatedFrom: [],
        id: "1",
        kind: "scriptSource",
        prettyPrinted: "pp1",
        prettyPrintedFrom: undefined,
        url: "/index.js",
      },
      pp1: {
        canonicalId: "1",
        contentHash: "contentHash#1",
        correspondingSourceIds: ["pp1"],
        generated: [],
        generatedFrom: [],
        id: "pp1",
        kind: "prettyPrinted",
        prettyPrinted: undefined,
        prettyPrintedFrom: "1",
        url: "/src/index.js",
      },
    });
  });

  it("can combine original, generated, and pretty-printed sources", () => {
    expect(
      newSourcesToCompleteSourceDetails([
        {
          contentHash: "contentHash#1",
          kind: "scriptSource",
          sourceId: "1",
          url: "/index.js",
        },
        {
          contentHash: "contentHash#o1",
          generatedSourceIds: ["1"],
          kind: "sourceMapped",
          sourceId: "o1",
          url: "/src/index.ts",
        },
        {
          generatedSourceIds: ["1"],
          kind: "prettyPrinted",
          sourceId: "pp1",
          url: "/src/index.js",
        },
        {
          generatedSourceIds: ["o1"],
          kind: "prettyPrinted",
          sourceId: "ppo1",
          url: "/src/index.ts",
        },
      ])
    ).toEqual({
      "1": {
        canonicalId: "o1",
        contentHash: "contentHash#1",
        correspondingSourceIds: ["1"],
        generated: [],
        generatedFrom: ["o1"],
        id: "1",
        kind: "scriptSource",
        prettyPrinted: "pp1",
        prettyPrintedFrom: undefined,
        url: "/index.js",
      },
      o1: {
        canonicalId: "o1",
        contentHash: "contentHash#o1",
        correspondingSourceIds: ["o1"],
        generated: ["1"],
        generatedFrom: [],
        id: "o1",
        kind: "sourceMapped",
        prettyPrinted: "ppo1",
        prettyPrintedFrom: undefined,
        url: "/src/index.ts",
      },
      pp1: {
        canonicalId: "o1",
        contentHash: "contentHash#1",
        correspondingSourceIds: ["pp1"],
        generated: [],
        generatedFrom: [],
        id: "pp1",
        kind: "prettyPrinted",
        prettyPrinted: undefined,
        prettyPrintedFrom: "1",
        url: "/src/index.js",
      },
      ppo1: {
        canonicalId: "o1",
        contentHash: "contentHash#o1",
        correspondingSourceIds: ["ppo1"],
        generated: [],
        generatedFrom: [],
        id: "ppo1",
        kind: "prettyPrinted",
        prettyPrinted: undefined,
        prettyPrintedFrom: "o1",
        url: "/src/index.ts",
      },
    });
  });

  it("can put together HTML sources and their extracted scripts", () => {
    expect(
      newSourcesToCompleteSourceDetails([
        {
          contentHash: "contentHash#h1",
          kind: "html",
          sourceId: "h1",
          generatedSourceIds: ["2"],
          url: "/index.html",
        },
        {
          contentHash: "contentHash#2",
          kind: "inlineScript",
          sourceId: "2",
          url: "/index.html",
        },
      ])
    ).toEqual({
      "2": {
        canonicalId: "h1",
        contentHash: "contentHash#2",
        correspondingSourceIds: ["2"],
        generated: [],
        generatedFrom: ["h1"],
        id: "2",
        kind: "inlineScript",
        prettyPrinted: undefined,
        prettyPrintedFrom: undefined,
        url: "/index.html",
      },
      h1: {
        canonicalId: "h1",
        contentHash: "contentHash#h1",
        correspondingSourceIds: ["h1"],
        generated: ["2"],
        generatedFrom: [],
        id: "h1",
        kind: "html",
        prettyPrinted: undefined,
        prettyPrintedFrom: undefined,
        url: "/index.html",
      },
    });
  });
});

it("can link corresponding sources", () => {
  expect(
    newSourcesToCompleteSourceDetails([
      {
        contentHash: "contentHash",
        kind: "html",
        sourceId: "h1",
        url: "/index.html",
      },
      {
        contentHash: "contentHash",
        kind: "html",
        sourceId: "h2",
        url: "/index.html",
      },
    ])
  ).toEqual({
    h1: {
      canonicalId: "h1",
      contentHash: "contentHash",
      correspondingSourceIds: ["h1", "h2"],
      generated: [],
      generatedFrom: [],
      id: "h1",
      kind: "html",
      prettyPrinted: undefined,
      prettyPrintedFrom: undefined,
      url: "/index.html",
    },
    h2: {
      canonicalId: "h2",
      contentHash: "contentHash",
      correspondingSourceIds: ["h1", "h2"],
      generated: [],
      generatedFrom: [],
      id: "h2",
      kind: "html",
      prettyPrinted: undefined,
      prettyPrintedFrom: undefined,
      url: "/index.html",
    },
  });
});

it("is not fooled by pretty-printed sources with missing contentHashes", () => {
  expect(
    newSourcesToCompleteSourceDetails([
      {
        contentHash: "contentHash#h1",
        kind: "html",
        sourceId: "h1",
        url: "/index.html",
      },
      {
        generatedSourceIds: ["h1"],
        kind: "prettyPrinted",
        sourceId: "pph1",
        url: "/index.html",
      },
      {
        contentHash: "contentHash#h2",
        kind: "html",
        sourceId: "h2",
        url: "/index.html",
      },
      {
        kind: "prettyPrinted",
        generatedSourceIds: ["h2"],
        sourceId: "pph2",
        url: "/index.html",
      },
    ])
  ).toEqual({
    h1: {
      canonicalId: "h1",
      contentHash: "contentHash#h1",
      correspondingSourceIds: ["h1"],
      generated: [],
      generatedFrom: [],
      id: "h1",
      kind: "html",
      prettyPrinted: "pph1",
      prettyPrintedFrom: undefined,
      url: "/index.html",
    },
    h2: {
      canonicalId: "h2",
      contentHash: "contentHash#h2",
      correspondingSourceIds: ["h2"],
      generated: [],
      generatedFrom: [],
      id: "h2",
      kind: "html",
      prettyPrinted: "pph2",
      prettyPrintedFrom: undefined,
      url: "/index.html",
    },
    pph1: {
      canonicalId: "h1",
      contentHash: "contentHash#h1",
      correspondingSourceIds: ["pph1"],
      generated: [],
      generatedFrom: [],
      id: "pph1",
      kind: "prettyPrinted",
      prettyPrinted: undefined,
      prettyPrintedFrom: "h1",
      url: "/index.html",
    },
    pph2: {
      canonicalId: "h2",
      contentHash: "contentHash#h2",
      correspondingSourceIds: ["pph2"],
      generated: [],
      generatedFrom: [],
      id: "pph2",
      kind: "prettyPrinted",

      prettyPrinted: undefined,
      prettyPrintedFrom: "h2",
      url: "/index.html",
    },
  });
});

describe("ThreadFront source methods", () => {
  beforeAll(() => {
    (
      [
        {
          contentHash: "contentHash#1",
          kind: "scriptSource",
          sourceId: "1",
          url: "/index.js",
        },
        {
          contentHash: "contentHash",
          kind: "html",
          sourceId: "h1",
          generatedSourceIds: ["2"],
          url: "/index.html",
        },
        {
          contentHash: "contentHash",
          kind: "html",
          sourceId: "h2",
          url: "/index.html",
        },
        {
          contentHash: "contentHash#2",
          kind: "inlineScript",
          sourceId: "2",
          url: "/index.html",
        },
      ] as newSource[]
    ).forEach(s => ThreadFront._newSourceListener(s));
    ThreadFront.hasAllSources = true;
    ThreadFront.groupSourceIds();
  });

  describe("getSourceIdsForUrl", () => {
    it("returns original versions of sources for the given URL", () => {
      expect(ThreadFront.getSourceIdsForURL("/index.js")).toEqual(["1"]);
      expect(ThreadFront.getSourceIdsForURL("/index.html")).toEqual(["h1", "h2"]);
    });
  });

  describe("getChosenSourceIdsForUrl", () => {
    it("returns the preferred and alternate sourceIds", () => {
      ThreadFront.preferSource("2", true);
      expect(ThreadFront.getChosenSourceIdsForUrl("/index.js")).toEqual([{ sourceId: "1" }]);
      expect(ThreadFront.getChosenSourceIdsForUrl("/index.html")).toEqual([
        { sourceId: "h1" },
        { sourceId: "h2" },
      ]);
      expect(ThreadFront.getChosenSourceIdsForUrl("/index.js")).toEqual([{ sourceId: "1" }]);
    });
  });

  describe("getCorrespondingSourceIds", () => {
    it("returns the preferred and alternate sourceIds", () => {
      expect(ThreadFront.getCorrespondingSourceIds("1")).toEqual(["1"]);
      expect(ThreadFront.getCorrespondingSourceIds("h1")).toEqual(["h1", "h2"]);
    });
  });

  describe("getSourceURLRaw", () => {
    it("returns the URL of the source", () => {
      expect(ThreadFront.getSourceURLRaw("1")).toEqual("/index.js");
    });
  });

  describe("getSourceURL", () => {
    it("returns the URL of the source", () => {
      expect(ThreadFront.getSourceURLRaw("1")).toEqual("/index.js");
    });
  });

  describe("getGeneratedSourceIdsForURL", () => {
    // This one seems odd to me?
    it("returns the URL of the generated sources", () => {
      expect(ThreadFront.getGeneratedSourceIdsForURL("/index.js")).toEqual(["1"]);
      expect(ThreadFront.getGeneratedSourceIdsForURL("/index.html")).toEqual(["h2", "2"]);
    });
  });

  describe("getGeneratedSourceIds", () => {
    // Why doesn't this line up with the getGeneratedSourceIdsForURL at all?
    // Also, it just so happens that this is only used by SmartTrace which is
    // only used by the console and object inspector, so if we rip out old
    // console code this code will be dead.
    it("returns the URL of the generated sources", () => {
      expect(ThreadFront.getGeneratedSourceIds("1")).toEqual(undefined);
      expect(ThreadFront.getGeneratedSourceIds("2")).toEqual(undefined);
      expect(ThreadFront.getGeneratedSourceIds("h1")).toEqual(["2"]);
    });
  });

  describe("getGeneratedLocation", () => {
    const location = (sourceId: string) => {
      return {
        sourceId,
        line: 1,
        column: 1,
      };
    };
    it("returns the generated location?", () => {
      // This method assumes that if any source has no "generatedSourceIds"
      // attached then *it* must be a generated location. Which is only true if
      // you assume that this is a set of locations with exactly one generated
      // source I think? Checking `generatedFrom` would be better, but maybe we
      // can find a more exact method to solve the problems this is currently
      // solving.
      // Only used by
      // packages/protocol/thread/value.ts:387
      expect(
        ThreadFront.getGeneratedLocation([location("1"), location("h2"), location("2")])
      ).toEqual(location("1"));
    });
  });
});
