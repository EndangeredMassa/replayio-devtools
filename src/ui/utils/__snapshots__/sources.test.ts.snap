// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can link corresponding sources 1`] = `
Object {
  "h1": Object {
    "canonicalId": "h1",
    "contentHash": "contentHash",
    "correspondingSourceIds": Array [
      "h1",
      "h2",
    ],
    "generated": Array [],
    "generatedFrom": Array [],
    "id": "h1",
    "kind": "html",
    "prettyPrinted": undefined,
    "prettyPrintedFrom": undefined,
    "url": "/index.html",
  },
  "h2": Object {
    "canonicalId": "h2",
    "contentHash": "contentHash",
    "correspondingSourceIds": Array [
      "h1",
      "h2",
    ],
    "generated": Array [],
    "generatedFrom": Array [],
    "id": "h2",
    "kind": "html",
    "prettyPrinted": undefined,
    "prettyPrintedFrom": undefined,
    "url": "/index.html",
  },
}
`;

exports[`is not fooled by pretty-printed sources with missing contentHashes 1`] = `
Object {
  "h1": Object {
    "canonicalId": "h1",
    "contentHash": "contentHash#h1",
    "correspondingSourceIds": Array [
      "h1",
    ],
    "generated": Array [],
    "generatedFrom": Array [],
    "id": "h1",
    "kind": "html",
    "prettyPrinted": "pph1",
    "prettyPrintedFrom": undefined,
    "url": "/index.html",
  },
  "h2": Object {
    "canonicalId": "h2",
    "contentHash": "contentHash#h2",
    "correspondingSourceIds": Array [
      "h2",
    ],
    "generated": Array [],
    "generatedFrom": Array [],
    "id": "h2",
    "kind": "html",
    "prettyPrinted": "pph2",
    "prettyPrintedFrom": undefined,
    "url": "/index.html",
  },
  "pph1": Object {
    "canonicalId": "h1",
    "contentHash": "contentHash#h1",
    "correspondingSourceIds": Array [
      "pph1",
    ],
    "generated": Array [],
    "generatedFrom": Array [],
    "id": "pph1",
    "kind": "prettyPrinted",
    "prettyPrinted": undefined,
    "prettyPrintedFrom": "h1",
    "url": "/index.html",
  },
  "pph2": Object {
    "canonicalId": "h2",
    "contentHash": "contentHash#h2",
    "correspondingSourceIds": Array [
      "pph2",
    ],
    "generated": Array [],
    "generatedFrom": Array [],
    "id": "pph2",
    "kind": "prettyPrinted",
    "prettyPrinted": undefined,
    "prettyPrintedFrom": "h2",
    "url": "/index.html",
  },
}
`;

exports[`newSourcesToCompleteSourceDetails can backlink generated sources properly 1`] = `
Object {
  "1": Object {
    "canonicalId": "o1",
    "contentHash": "contentHash#1",
    "correspondingSourceIds": Array [
      "1",
    ],
    "generated": Array [],
    "generatedFrom": Array [
      "o1",
    ],
    "id": "1",
    "kind": "scriptSource",
    "prettyPrinted": undefined,
    "prettyPrintedFrom": undefined,
    "url": "/index.js",
  },
  "o1": Object {
    "canonicalId": "o1",
    "contentHash": "contentHash#o1",
    "correspondingSourceIds": Array [
      "o1",
    ],
    "generated": Array [
      "1",
    ],
    "generatedFrom": Array [],
    "id": "o1",
    "kind": "sourceMapped",
    "prettyPrinted": undefined,
    "prettyPrintedFrom": undefined,
    "url": "/index.ts",
  },
}
`;

exports[`newSourcesToCompleteSourceDetails can combine original, generated, and pretty-printed sources 1`] = `
Object {
  "1": Object {
    "canonicalId": "o1",
    "contentHash": "contentHash#1",
    "correspondingSourceIds": Array [
      "1",
    ],
    "generated": Array [],
    "generatedFrom": Array [
      "o1",
    ],
    "id": "1",
    "kind": "scriptSource",
    "prettyPrinted": "pp1",
    "prettyPrintedFrom": undefined,
    "url": "/index.js",
  },
  "o1": Object {
    "canonicalId": "o1",
    "contentHash": "contentHash#o1",
    "correspondingSourceIds": Array [
      "o1",
    ],
    "generated": Array [
      "1",
    ],
    "generatedFrom": Array [],
    "id": "o1",
    "kind": "sourceMapped",
    "prettyPrinted": "ppo1",
    "prettyPrintedFrom": undefined,
    "url": "/src/index.ts",
  },
  "pp1": Object {
    "canonicalId": "o1",
    "contentHash": "contentHash#1",
    "correspondingSourceIds": Array [
      "pp1",
    ],
    "generated": Array [],
    "generatedFrom": Array [],
    "id": "pp1",
    "kind": "prettyPrinted",
    "prettyPrinted": undefined,
    "prettyPrintedFrom": "1",
    "url": "/src/index.js",
  },
  "ppo1": Object {
    "canonicalId": "o1",
    "contentHash": "contentHash#o1",
    "correspondingSourceIds": Array [
      "ppo1",
    ],
    "generated": Array [],
    "generatedFrom": Array [],
    "id": "ppo1",
    "kind": "prettyPrinted",
    "prettyPrinted": undefined,
    "prettyPrintedFrom": "o1",
    "url": "/src/index.ts",
  },
}
`;

exports[`newSourcesToCompleteSourceDetails can link pretty-printed and minified sources 1`] = `
Object {
  "1": Object {
    "canonicalId": "1",
    "contentHash": "contentHash#1",
    "correspondingSourceIds": Array [
      "1",
    ],
    "generated": Array [],
    "generatedFrom": Array [],
    "id": "1",
    "kind": "scriptSource",
    "prettyPrinted": "pp1",
    "prettyPrintedFrom": undefined,
    "url": "/index.js",
  },
  "pp1": Object {
    "canonicalId": "1",
    "contentHash": "contentHash#1",
    "correspondingSourceIds": Array [
      "pp1",
    ],
    "generated": Array [],
    "generatedFrom": Array [],
    "id": "pp1",
    "kind": "prettyPrinted",
    "prettyPrinted": undefined,
    "prettyPrintedFrom": "1",
    "url": "/src/index.js",
  },
}
`;

exports[`newSourcesToCompleteSourceDetails can put together HTML sources and their extracted scripts 1`] = `
Object {
  "2": Object {
    "canonicalId": "h1",
    "contentHash": "contentHash#2",
    "correspondingSourceIds": Array [
      "2",
    ],
    "generated": Array [],
    "generatedFrom": Array [
      "h1",
    ],
    "id": "2",
    "kind": "inlineScript",
    "prettyPrinted": undefined,
    "prettyPrintedFrom": undefined,
    "url": "/index.html",
  },
  "h1": Object {
    "canonicalId": "h1",
    "contentHash": "contentHash#h1",
    "correspondingSourceIds": Array [
      "h1",
    ],
    "generated": Array [
      "2",
    ],
    "generatedFrom": Array [],
    "id": "h1",
    "kind": "html",
    "prettyPrinted": undefined,
    "prettyPrintedFrom": undefined,
    "url": "/index.html",
  },
}
`;

exports[`newSourcesToCompleteSourceDetails should be able to return a complete source from a single newSource event 1`] = `
Object {
  "1": Object {
    "canonicalId": "1",
    "contentHash": "contentHash#1",
    "correspondingSourceIds": Array [
      "1",
    ],
    "generated": Array [],
    "generatedFrom": Array [],
    "id": "1",
    "kind": "scriptSource",
    "prettyPrinted": undefined,
    "prettyPrintedFrom": undefined,
    "url": "/index.js",
  },
}
`;
